### PR TITLE
chore(participant) Wire up telemetry for user prompts VSCODE-606

### PR DESCRIPTION
**Pull request by nirinchev**
## Description
Adds some telemetry stats when constructing prompts. Part of VSCODE-575.

### Checklist
- [ ] New tests and/or benchmarks are included
- [ ] Documentation is changed or added
- [ ] I have signed the MongoDB Contributor License Agreement (https://www.mongodb.com/legal/contributor-agreement)

## Motivation and Context

- [ ] Bugfix
- [ ] New feature
- [ ] Dependency update
- [ ] Misc

## Types of changes

- [ ] *Backport Needed*
- [ ] Patch (non-breaking change which fixes an issue)
- [ ] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: src/telemetry/telemetryService.ts [hunk 1]
```diff
@@ -108,6 +108,13 @@ type ParticipantResponseFailedProperties = {
   error_name: ParticipantErrorTypes;
 };
 
+export type ParticipantPromptProperties = {
+  command: string;
+  user_input_length: number;
+  total_message_length: number;
+  has_sample_documents: boolean;
```
Comment (Anemy) [inline src/telemetry/telemetryService.ts:115]: Should we also track the amount of messages in the chat? Could be useful to see how much people start new chats vs keep existing ones going.

File: src/participant/participant.ts [hunk 2]
```diff
@@ -164,10 +165,10 @@ export default class ParticipantController {
   }
 
   async _getChatResponse({
-    messages,
+    prompt,
     token,
   }: {
-    messages: vscode.LanguageModelChatMessage[];
+    prompt: PromptResult;
```
Comment (Anemy) [inline src/participant/participant.ts:171]: [nit] No action needed here, mostly thinking out loud. I wonder if we're over using the term `prompt` here. (or maybe it's in other places we could change it).
Is there a way we can name this to get the meaning quicker on first read? Maybe `ChatRequest`? `ChatCompletionRequest`?

File: src/participant/participant.ts [hunk 3]
```diff
@@ -176,20 +177,22 @@ export default class ParticipantController {
       throw new Error('Copilot model not found');
     }
 
-    return await model.sendRequest(messages, {}, token);
+    this._telemetryService.trackCopilotParticipantPrompt(prompt.stats);
```
Comment (Anemy) [inline src/participant/participant.ts:180]: This is going to track for our metadata requests - `namespace` and `intent`. Should we want track those separately, or least add an indicator for them?

**Review by Anemy — COMMENTED** [review]
Nice test adds, a few questions on what we're tracking, I think we want to think about how we can distinguish the metadata prompt chaining requests from others.

File: src/participant/participant.ts (hunk 2 above)
Comment (alenakhineika) [reply src/participant/participant.ts:171]: Also, it is a bit hard to get while reading, that prompt is an input argument but has PromptResult type. Like input is a result.

File: src/participant/participant.ts (hunk 3 above)
Comment (nirinchev) [reply src/participant/participant.ts]: I've added `internal_purpose` to differentiate namespace and intent requests.

File: src/participant/participant.ts (hunk 2 above)
Comment (nirinchev) [reply src/participant/participant.ts]: Agree - wasn't quite happy with the name myself. I changed it to `ModelInput` - still don't think it's ideal, but I feel it's more descriptive than before.

File: src/telemetry/telemetryService.ts (hunk 1 above)
Comment (nirinchev) [reply src/telemetry/telemetryService.ts:117]: Added `history_length` for this. Note that this only tracks the number of history messages and not the length of the messages themselves.

**Review by Anemy — APPROVED** [review]

**Pull request merged**